File: test/commands/signature.test.js
import {
  runClangdQuery,
  assert,
  assertContains,
  assertMatches,
  countOccurrences,
  waitForDaemonReady,
} from '../test-helpers.js';

console.log('Testing signature command...\n');

// Ensure daemon is ready before running tests
console.log('Waiting for daemon to be ready...');
await waitForDaemonReady();

// Test 1: Get signature of a constructor
console.log('Test 1: Get signature of GameObject constructor');
const result1 = await runClangdQuery(['signature', 'GameObject']);
assert(result1.exitCode === 0, 'Command should succeed');
// Should show the constructor signature
assertContains(result1.stdout, 'GameObject::GameObject(const std::string &name)');
assertContains(result1.stdout, 'Parameters:');
assertContains(result1.stdout, 'const std::string & name');
assertContains(result1.stdout, 'Modifiers: explicit');
console.log('✓ Test 1 passed\n');

// Test 2: Get signature of a method
console.log('Test 2: Get signature of Update method');
const result2 = await runClangdQuery(['signature', 'Update']);
assert(result2.exitCode === 0, 'Command should succeed');
// Should show Update method signatures
assertContains(result2.stdout, 'Update');
assertContains(result2.stdout, 'Parameters:');
assertContains(result2.stdout, 'float delta_time');
console.log('✓ Test 2 passed\n');

// Test 3: Get signature of a template function
console.log('Test 3: Get signature of GetComponent template method');
const result3 = await runClangdQuery(['signature', 'GetComponent']);
assert(result3.exitCode === 0, 'Command should succeed');
// Should show the template method signature
assertContains(result3.stdout, 'GetComponent');
assertContains(result3.stdout, 'Template Parameters: <typename T>');
assertContains(result3.stdout, 'std::optional<std::shared_ptr<T>>');
console.log('✓ Test 3 passed\n');

// Test 4: Get signature of non-existent function
console.log('Test 4: Get signature of non-existent function');
const result4 = await runClangdQuery(['signature', 'NonExistentFunction']);
assert(result4.exitCode === 0, 'Command should succeed even with no results');
assertContains(result4.stdout, "No function or method named 'NonExistentFunction' found");
console.log('✓ Test 4 passed\n');

console.log('All signature tests passed! ✓');

File: go/test/signature_test.go
package test

import (
	"testing"
)

func TestSignatureCommand(t *testing.T) {
	tc := NewTestContext(t)
	
	// Ensure daemon is ready before running tests
	t.Log("Waiting for daemon to be ready...")
	tc.WaitForDaemonReady()
	
	t.Run("Get signature of GameObject constructor", func(t *testing.T) {
		result := tc.RunCommand("signature", "GameObject")
		tc.AssertExitCode(result, 0)
		// Should show the constructor signature
		tc.AssertContains(result.Stdout, "GameObject::GameObject(const std::string &name)")
		tc.AssertContains(result.Stdout, "Parameters:")
		tc.AssertContains(result.Stdout, "const std::string & name")
		tc.AssertContains(result.Stdout, "Modifiers: explicit")
	})
	
	t.Run("Get signature of Update method", func(t *testing.T) {
		result := tc.RunCommand("signature", "Update")
		tc.AssertExitCode(result, 0)
		// Should show Update method signatures
		tc.AssertContains(result.Stdout, "Update")
		tc.AssertContains(result.Stdout, "Parameters:")
		tc.AssertContains(result.Stdout, "float delta_time")
	})
	
	t.Run("Get signature of GetComponent template method", func(t *testing.T) {
		result := tc.RunCommand("signature", "GetComponent")
		tc.AssertExitCode(result, 0)
		// Should show the template method signature
		tc.AssertContains(result.Stdout, "GetComponent")
		tc.AssertContains(result.Stdout, "Template Parameters: <typename T>")
		tc.AssertContains(result.Stdout, "std::optional<std::shared_ptr<T>>")
	})
	
	t.Run("Get signature of non-existent function", func(t *testing.T) {
		result := tc.RunCommand("signature", "NonExistentFunction")
		tc.AssertExitCode(result, 0)
		tc.AssertContains(result.Stdout, "No function or method named 'NonExistentFunction' found")
	})
	
	// Additional test: overloaded methods
	t.Run("Get signature of overloaded method", func(t *testing.T) {
		result := tc.RunCommand("signature", "Initialize")
		tc.AssertExitCode(result, 0)
		// Should show Initialize method signature
		tc.AssertContains(result.Stdout, "Initialize")
		tc.AssertContains(result.Stdout, "Parameters:")
	})
}